presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java

```diff
@@ -285,9 +286,13 @@ public PlanNode replaceChildren(List<PlanNode> newChildren)
         PlanNode newLeft = newChildren.get(0);
         PlanNode newRight = newChildren.get(1);
         // Reshuffle join output symbols (for cross joins) since order of symbols in child nodes might have changed
-        List<Symbol> newOutputSymbols = Stream.concat(newLeft.getOutputSymbols().stream(), newRight.getOutputSymbols().stream())
-                .filter(outputSymbols::contains)
-                .collect(toImmutableList());
+        List<Symbol> newOutputSymbols = outputSymbols;
+        if (isCrossJoin()) {
+            Set<Symbol> outputSymbolSet = new HashSet<>(outputSymbols);
+            newOutputSymbols = Stream.concat(newLeft.getOutputSymbols().stream(), newRight.getOutputSymbols().stream())
+                    .filter(outputSymbolSet::contains)
+                    .collect(toImmutableList());
+        }
```

Comment:
+1

Reply:
Okay. But
>We don't want to produce symbols that were already pruned

But for this operation we require that check right ? Correct me If I am wrong

Reply:
> But for this operation we require that check right ? Correct me If I am wrong

I think that replaced children shouldn't produce different set of symbols actually. They can produce different order of symbols though. 

Since symbols are same, we don't have to touch `outputSymbols` at all (if we don't care about reshuffle)

Reply:
>replaced children shouldn't produce different set of symbols actually

So for `PruneCrossJoinColumns` rule can we change it to create a new `JoinNode` instead of replacing its children using this `restrictChildOutputs`.(`PruneCrossJoinColumns` violates this condition)

Reply:
I think so.

Reply:
> So for PruneCrossJoinColumns rule can we change it to create a new JoinNode instead of replacing its children using this restrictChildOutputs.(PruneCrossJoinColumns violates this condition)

I think pruning children symbols via `replaceChildren` should also work as long as we don't remove symbols used by `JoinNode.outputSymbols`.

Consider adding `checkState` in `com.facebook.presto.sql.planner.plan.JoinNode#replaceChildren`:
```
checkState(union(
  ImmutableSet.of(newLeft.getOutputSymbols()), 
  ImmutableSet.of(newRight.getOutputSymbols())).containsAll(outputSymbols), ...)
```

Reply:
But for a `CrossJoin` it's `output symbol = left node's symbol + right node's symbol` (order is not considered) so if there is a change in any of its children (left or right) it would reflect on `JoinNode.outputSymbol`. 

>Consider adding checkState in com.facebook.presto.sql.planner.plan.JoinNode#replaceChildren:

This is already present in `JoinNode` constructor so do we need to add it here also ?

Reply:
> This is already present in JoinNode constructor so do we need to add it here also ?

Oki. Then its not needed.